videostore/mux.go

```diff
@@ -0,0 +1,559 @@
+package videostore
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/bluenviron/mediacommon/pkg/codecs/h264"
+	"github.com/bluenviron/mediacommon/pkg/codecs/h265"
+	"github.com/viam-modules/viamrtsp/registry"
+	"github.com/viam-modules/video-store/videostore"
+	"go.viam.com/rdk/logging"
+	"go.viam.com/rdk/resource"
+	"go.viam.com/utils"
+)
+
+const monitorInterval = time.Second * 5
+
+type metadata struct {
+	width        int
+	height       int
+	vps          []byte
+	sps          []byte
+	pps          []byte
+	dtsExtractor interface {
+		Extract(au [][]byte, pts int64) (int64, error)
+	}
+	spsUnChanged       bool
+	firstTimeStampsSet bool
+	firstPTS           int64
+	firstDTS           int64
+}
+type rawSegmenterMux struct {
+	// are valid for the lifetime of the rawSegmenterMux
+	camName resource.Name
+	logger  logging.Logger
+	worker  *utils.StoppableWorkers
+	regDone <-chan struct{}
+	cam     registry.ModuleCamera
+
+	mu       sync.Mutex
+	rawSeg   *videostore.RawSegmenter
+	codec    atomic.Int64
+	metadata metadata
+}
+
+var codecs = []videostore.CodecType{
+	videostore.CodecTypeH265,
+	videostore.CodecTypeH264,
+}
+
+func newRawSegmenterMux(rawSeg *videostore.RawSegmenter, camName resource.Name, logger logging.Logger) *rawSegmenterMux {
+	return &rawSegmenterMux{
+		rawSeg:  rawSeg,
+		camName: camName,
+		worker:  utils.NewBackgroundStoppableWorkers(),
+		logger:  logger,
+	}
+}
+
+// init and close are called by videostore.
+func (m *rawSegmenterMux) init() error {
+	if m.rawSeg == nil {
+		return errors.New("videostore.RTPVideoStore.Segmenter() is nil")
+	}
+	cam, err := registry.Global.Get(m.camName.String())
```

Comment:
Trying to understand the flow better here. Is there any possibility of race condition in`registry.Global` reference between vaimrtrtsp and video store spin up? Is it guaranteed that the camera will be registered before the muxer attempts to reference it here?

Reply:
I don't think so.

If the video-store's config has a camera then the video-store's config will return that camera as a dependency of video-store. Resource graph will not boot video-store until the camera is booted.